src/pp.rs

```diff
@@ -368,45 +375,13 @@ impl<'a> DirectiveProcessor<'a> {
         Ok(())
     }
 
-    fn evaluate_if_expression(&mut self, location: Location, line: Vec<Token>, max_recursion_depth: usize) -> Step<bool> {
-        let line: Vec<_> = line.into_iter().map(|t| t.value).collect();
-
-        match *line {
-            [TokenValue::Integer(ref i)] => {
-                // Evaluate "#if 123"
-                Ok(i.value != 0)
-            }
-            [TokenValue::Ident(ref ident), TokenValue::Punct(Punct::LeftParen), 
-                TokenValue::Ident(ref ident2), TokenValue::Punct(Punct::RightParen)] |
-            [TokenValue::Ident(ref ident), TokenValue::Ident(ref ident2)] if ident == "defined" => {
-                // Evaluate "#if defined XYZ" and "#if defined(XYZ)"
-                Ok(self.defines.contains_key(ident2))
-            }
-            [TokenValue::Punct(Punct::Bang), TokenValue::Ident(ref ident), TokenValue::Punct(Punct::LeftParen), 
-                TokenValue::Ident(ref ident2), TokenValue::Punct(Punct::RightParen)] |
-            [TokenValue::Punct(Punct::Bang), TokenValue::Ident(ref ident), TokenValue::Ident(ref ident2)] if ident == "defined" => {
-                // Evaluate "#if !defined XYZ" and "#if !defined(XYZ)"
-                Ok(!self.defines.contains_key(ident2))
-            }
-            [TokenValue::Ident(ref ident)] => {
-                // Handle "#if XYZ"
-                if max_recursion_depth == 0 {
-                    return Err(StepExit::Error((PreprocessorError::RecursionLimitReached, location)));
-                }
-                match self.defines.get(ident) {
-                    Some(d) => {
-                        assert_eq!(d.tokens.len(), 1); // TODO
-                        let line = d.tokens.clone();
-                        self.evaluate_if_expression(location, line, max_recursion_depth - 1)
-                    }
-                    None => Err(StepExit::Error((PreprocessorError::MacroNotDefined, location))),
-                }
-            }
-            _ => {
-                // TODO, so much to do here xD
-                todo!();
-            }
-        }
+    fn evaluate_if_expression(
+        &mut self,
+        location: Location,
+        line: Vec<Token>,
+        max_recursion_depth: usize,
+    ) -> Step<bool> {
+        if_parser::IfParser::new(line, max_recursion_depth, &self.defines, location).evaluate()
```

Comment:
I think that with this change the max_recursion_depth can be removed as it was only used previously to prevent infinite macro expansion in the special-cased evaluate_if_expression.

Reply:
I've finished implementing using the `MELexer` and removed `max_recursion_depth` but now there's a test like this

```rust
    check_preprocessing_error(
        "#define FOO FOO
         #if FOO
         #endif",
        PreprocessorError::RecursionLimitReached,
    );
```

It's failing with `EndOfInput` what do you want me to do with the test

Reply:
That GLSL code should still be invalid, so it should be enough to just change `PreprocessorError::RecursionLimitReached` to the error your version produces?

Reply:
I think right now it should do the following steps: `#if FOO`, expand `FOO`, produce token `Ident(Foo)` (since a macro can only be expanding once at a time), so it becomes `#if Ident(Foo)` that is `#if 0`.

I'm not sure what is producing the `EndOfInput`. @JCapucho, would it be possible to update this PR with the WIP code you have (or open a different one) so we can take a look and debug with you?

Reply:
`#if Ident(Foo)` expands to `#if 0` when Foo isn't defined? I would have expected that to just produce an error about Foo not being defined.

Reply:
Yeah it's weird but what all C compilers do. (I haven't checked ANGLE though)

Reply:
No the c spec explicitly mentions this behaviour only for `#if`.

I'm not home right now and probably won't be until tomorrow but I'll push my current changes as soon as I can.

Reply:
I've fixed the issue it was a dumb mistake on my part